Add API docs tooling

kdmccormick opened the pull request:
@nedbat @regisb -- Here's my stab at pulling the API doc tools out of `edx-platform` and into a shared repository. I settled on `edx-drf-extensions` because it's our canonical "REST API tooling" library and is already installed into most edX IDAs.

This PR creates a new folder called `edx_api_doc_tools` that exposes a toolkit of utilities through its `__init__.py`. It also has a silly "edX Hedgehog API" under `edx_api_doc_tools/example` that demonstrates how to use the doc tools. Finally, `edx_api_doc_tools/tests` does some baseline testing by looking at the JSON generated by from the example API.

There are a few `TODO`s scattered throughout this PR (hence its DRAFT status), but I wanted to solicit feedback early in case there are any major changes I need to make.

I haven't included Sphinx doc generation in this PR; I'm thinking that it could merge separately, while applying this PR's changes to `edx-platform` could be done in parallel.

Lastly, @regis and @nedbat , I plan on giving you both co-authorship on this final commit(s), as you wrote much of the original code; let me know if you have any qualms with that :)

nedbat commented on edx_api_doc_tools/conf_utils.py:
```diff
@@ -0,0 +1,148 @@
+
+def make_docs_urls(api_info):
+    """
+    Convenience function for creating API doc views given an API info object.
```
This docstring could use an example of how this function is used, how is it helpful? Ditto for the other functions in this module.

kdmccormick replied:
Added some examples.

nedbat commented on edx_api_doc_tools/conf_utils.py:
```diff
@@ -0,0 +1,148 @@
+    return openapi.Info(
+        title=title,
+        default_version=version,
+        contact=openapi.Contact(email=email or "ocsm@edx.org"),
```
No need to default the email twice.

nedbat commented on edx_api_doc_tools/data.py:
```diff
@@ -0,0 +1,100 @@
+OpenAPI schema types. We wrap them so that we can
+easily switch from drf-yasg to another supporting
+library if we must without affecting clients of this
+library.
```
Why do you wrap docstrings and comments so narrowly?

kdmccormick replied:
Bad habit, or an overreaction to seeing 120+-wide unwrapped docstrings, I guess. I'll try hard to wrap at column ~80.

nedbat commented on edx_api_doc_tools/example/test_example.py:
```diff
@@ -0,0 +1,86 @@
+
+    def test_post(self):
+        response = self.request('post', 'hogs/', self.new_hog)
+        assert response.status_code == 501
```
Why is this 501?

nedbat replied:
Oh: because the method raises NotImplementedSomethingOrOther.

nedbat commented on edx_api_doc_tools/example/views.py:
```diff
@@ -0,0 +1,102 @@
+@schema_for(
+    'retrieve',
+    # TODO: The below string becomes the operation's description instead
+    #       of its summary for some reason.
```
Something wrong with the multi-line decision making?

kdmccormick replied:
Yeah, I'm debugging it currently.

nedbat commented on edx_api_doc_tools/example/views.py:
```diff
@@ -0,0 +1,102 @@
+
+    def perform_create(self, serializer):
+        """
+        TODO
```
Do we have to implement these? Just for a test API, do we need those methods? Seems like that would be testing DRF, not our code.

nedbat commented on edx_api_doc_tools/tests/test_doc_tools.py:
```diff
@@ -0,0 +1,64 @@
+    * Make sure the browsable API looks correct,
+      and make sure the contents of the schema file look sane.
+    * If everything looks right, copy actual_schema.json into expected_schema.json
+      and commit the change, making it the new reference schema.
```
I like these instructions :)

nedbat commented on edx_api_doc_tools/tests/test_doc_tools.py:
```diff
@@ -0,0 +1,64 @@
+    def test_get_ui_view(self):
+        response = self.client.get('/api-docs/')
+        assert response.status_code == 200
+        assert 'edX Hedgehog Service API' in response.content.decode('utf-8')
```
You mention wanting to beef this up: I think asserting about a handful of other strings will be good.

kdmccormick replied:
@nedbat Realizing now that this is the maximal thing I can assert, and it's passing because that string is actually the `<title>`. The API spec is loaded via an AJAX call, which populates all page content.

nedbat commented on edx_api_doc_tools/tests/test_doc_tools.py:
```diff
@@ -0,0 +1,64 @@
+    path_of_actual_schema = join(base_path, 'actual_schema.json')
+
+    @classmethod
+    def setUpClass(cls):
```
setUpClass seems overkill here, since you only use the expected schema in one test. Maybe you had thought there would be more?

kdmccormick replied:
Yep, that was it. I'm going to break this into function tests.

nedbat commented on edx_api_doc_tools/view_utils.py:
```diff
@@ -0,0 +1,125 @@
+        'post', 'POST', 'create',
+        'put', 'PUT',
+        'patch', 'PATCH', 'update',
+        'delete', 'DELETE', and 'destroy'.
```
Why does this decorator care what the method name is?  Maybe this should say, "DRF method names include:"

kdmccormick replied:
It doesn't, but `swagger_auto_schema` only works on those names. Regardless, I like your rewording.

nedbat commented on edx_api_doc_tools/view_utils.py:
```diff
@@ -0,0 +1,125 @@
+        return responses
+    return {
+        http_status: (
+            str(value)
```
This will fail for non-ascii data, right? Shouldn't it be `value.encode("utf-8")` ?

kdmccormick commented on edx_rest_framework_extensions/__init__.py:
```diff
@@ -1,3 +1,3 @@
 """ edx Django REST Framework extensions. """
 
-__version__ = '2.4.6'  # pragma: no cover
+__version__ = '3.0.0'  # pragma: no cover
```
There was none, but I propose bumping it anyway, for no other reason than it'd be easy to remember that "Version 3 has the doc tools"

kdmccormick replied:
"Version 2.5 has the doc tools" is already fairly memorable, I guess.

kdmccormick replied:
(moot, because we're moving over to edx-api-doc-tools)

regisb commented:
Hey @kdmccormick! Good job working on this. The tests did not exist before in edx-platform, right? I can see that a lot of effort has gone into this.
I have qualms about the integration with the edx-drf-extensions repo: 

1. The large number of changes made to the requirements files show that the edx_api_doc_tools don't have much in common with the rest of this repo.
2. Unless I'm mistaken, edx_api_doc_tools does not depend on the rest of the code in the repo. Coalescing both repos would introduce an unnecessary dependency.
3. This PR adds 1k+ lines of code to a repo that has just 2.3k lines of code (roughly). This seems disproportionate.

Also, we should try to make it as easy as possible to use the API documentation tools. For that, we need to provide some documentation on how to use it. It would be great if we could have a guide on how to use the documentation tools as well as a reference of the various APIs. An additional benefit of this guide is that it would inform us of non backward-compatible changes.

nedbat commented:
@regisb We are definitely planning to thoroughly document the process of documenting APIs.  I guess @kdmccormick and I didn't let you know, but we have an ad-hoc to-do list in the old repo: https://github.com/edx/api-doc-tools/issues

I hear what you are saying about the api tooling perhaps being better in a separate repo.  I think the reasoning was that most repos that need edx-drf-extensions will also need the doc tooling, and vice-versa. Putting it all in one repo reduces the maintenance chores in the repos.  So it's a pragmatic choice rather than a philosophical one.  I could be swayed in the other direction.

regisb commented:
> I guess @kdmccormick and I didn't let you know, but we have an ad-hoc to-do list in the old repo: https://github.com/edx/api-doc-tools/issues

This is really cool! I wasn't aware that there was a dedicated repo.

> I think the reasoning was that most repos that need edx-drf-extensions will also need the doc tooling, and vice-versa.

The rub lies in the "most". If there is ever a non DRF-based API that comes out, it will have to install DRF nonetheless in order to document its API.

> Putting it all in one repo reduces the maintenance chores in the repos.

I disagree with this. For instance: if a repo depends on a feature included in release X+1 of api-doc-tools, but is not yet compatible with release X of edx-drf-extensions, then it will have to upgrade before it can start using the new feature.

> So it's a pragmatic choice rather than a philosophical one. I could be swayed in the other direction.

I'll do my best to convince you with practical arguments :)

kdmccormick commented:
@nedbat @regisb Thanks for all the feedback!

@nedbat I've applied some of your suggested changes; I'll apply the next week.

@regisb Re: `api-doc-tools` vs. `edx-drf-extensions`: You make good points. I hemmed and hawed between the two locations for a while. Within the walls of 141 Portland there seems to be a sentiment of "we have too many repositories", and every current edx.org IDA that needs API documentations also needs DRF extensions. However, I understand those two thing aren't necessarily true in the Open edX community. I'll stew on this and update you with my thoughts when I come back to this PR next week.

As a point of pedantry, the API tools are powered by `drf-yasg` which required DRF anyway.

Finally, regarding documentation on how to use these tools: that is absolutely part of the plan! My bad for omitting that from the description.

kdmccormick commented:
@regisb @nedbat Keeping this out of DRFX and making use of https://github.com/edx/api-doc-tools sounds good to me. I'll post a similar PR to that repo this week.

nedbat commented:
Weren't we going to finish the tweaks here before moving to the other repo?

kdmccormick commented:
@nedbat Right; that was the plan.

regisb commented:
> Keeping this out of DRFX and making use of https://github.com/edx/api-doc-tools sounds good to me. I'll post a similar PR to that repo this week.

Thanks! I understand the practical issues around the management of many different repos, and I appreciate the effort.

kdmccormick commented:
@nedbat @regisb I finished addressing the comments on this PR, and will be working off of https://github.com/edx/api-doc-tools/pull/13 from now on. I'll ping you both when I'm ready for another round of review.

kdmccormick commented on requirements/constraints.txt:
```diff
@@ -8,5 +8,8 @@
+backports.functools-lru-cache ; python_version == "2.7"
+functools32 ; python_version == "2.7"
 futures ; python_version == "2.7"
+ruamel.ordereddict ; python_version == "2.7"
```
@nedbat It's a transitive dependency through `ruamel.yaml`. It fails to install in Py3, which is why I added this constraint.
